Dockerfile

```diff
@@ -1,51 +1,87 @@
+ARG RUSTC_VERSION
+ARG ZIG_VERSION
+ARG TARGETPLATFORM
 
 # We need rustup so we have a sensible rust version, the version packed with bullsye is too old
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --profile minimal
+RUN curl --proto '=https' --tlsv1.2 -sSf  https://sh.rustup.rs | sh -s -- -y --default-toolchain "${RUSTC_VERSION}" --profile minimal
 ENV PATH="/root/.cargo/bin:${PATH}"
 
-# arm64 builds consume a lot of memory if `CARGO_NET_GIT_FETCH_WITH_CLI` is not
-# set to true, so we expose it as a build-arg.
-ARG CARGO_NET_GIT_FETCH_WITH_CLI=false
-ENV CARGO_NET_GIT_FETCH_WITH_CLI=$CARGO_NET_GIT_FETCH_WITH_CLI
+RUN rustup target add  \
+  --toolchain "${RUSTC_VERSION}" \
+  x86_64-unknown-linux-gnu \
+  aarch64-unknown-linux-gnu
+
+# Install zig and cargo-zigbuild, which are used by napi to cross-compile the native module
+RUN \
+  curl --proto '=https' --tlsv1.2 -sSf "https://ziglang.org/download/${ZIG_VERSION}/zig-linux-$(uname -m)-${ZIG_VERSION}.tar.xz" | tar -J -x -C /usr/local && \
+  ln -s "/usr/local/zig-linux-$(uname -m)-${ZIG_VERSION}/zig" /usr/local/bin/zig
 
-# Needed to build rust things for matrix-sdk-crypto-nodejs
-# See https://github.com/matrix-org/matrix-rust-sdk-bindings/blob/main/crates/matrix-sdk-crypto-nodejs/release/Dockerfile.linux#L5-L6
-RUN apt-get update && apt-get install -y build-essential cmake
+RUN cargo install --locked cargo-zigbuild@=${CARGO_ZIGBUILD_VERSION}
 
 WORKDIR /src
 
 COPY package.json yarn.lock ./
 RUN yarn config set yarn-offline-mirror /cache/yarn
-RUN yarn --ignore-scripts --pure-lockfile --network-timeout 600000
-
-COPY . ./
+RUN --mount=type=cache,target=/cache/yarn \
+    yarn --ignore-scripts --pure-lockfile --network-timeout 600000
 
 # Workaround: Need to install esbuild manually https://github.com/evanw/esbuild/issues/462#issuecomment-771328459
 RUN node node_modules/esbuild/install.js
-RUN yarn build
 
+COPY . ./
 
-# Stage 1: The actual container
-FROM node:18
+RUN --mount=type=cache,target=/cache/yarn \
+    sh ./scripts/docker-cross-env.sh "$TARGETPLATFORM" \
+    yarn build
 
-WORKDIR /bin/matrix-hookshot
 
-COPY --from=builder /src/yarn.lock /src/package.json ./
-COPY --from=builder /cache/yarn /cache/yarn
+# Stage 2: Install the production dependencies
+FROM --platform=${BUILDPLATFORM} node:${NODEJS_VERSION} AS deps
+
+ARG TARGETPLATFORM
+ARG MATRIX_SDK_VERSION
+
+WORKDIR /src
+
+COPY yarn.lock package.json scripts/docker-cross-env.sh scripts/docker-download-sdk.sh ./
 RUN yarn config set yarn-offline-mirror /cache/yarn
 
-RUN yarn --network-timeout 600000 --production --pure-lockfile && yarn cache clean
+RUN --mount=type=cache,target=/cache/yarn \
+    sh ./docker-cross-env.sh "$TARGETPLATFORM" \
+    yarn --ignore-scripts --pure-lockfile --network-timeout 600000 --production
+# Workaround: the install script of the matrix-rust-sdk only installs for the current platform, not the target one
+RUN sh ./docker-download-sdk.sh "$TARGETPLATFORM" "$MATRIX_SDK_VERSION"
```

Comment:
Did a PR for that: https://github.com/matrix-org/matrix-rust-sdk-crypto-nodejs/pull/8

Reply:
With that PR merged, would this line call `docker-cross-env.sh` instead?

Reply:
The `yarn install` above would be enough if we remove the `--ignore-scripts` flag

Reply:
I don't have write permissions on that repo so I can't merge myself
Besides, this would need a new release of the crypto SDK, which is probably not ready (especially since it moved repo)

Reply:
Ah, of course.  I'll get that kicked off then